Utxo coin lib

sedited opened the pull request.

sedited committed:
Basic implementation of typescript helper functions for bitcoinjs-lib

This commit implements some basic interfaces to be used in the future by
the edge wallet. They include bip32 key generation, transforms between
different key formats and transaction generation.

A lot of function signatures and interfaces currently contain spurious
definitions that lead to potentially undefined data. This is by
intention, since there is no error handling yet. Error definitions still
need to be defined properly.

The commit also does not implement multicoin support yet. Again, this
needs some further consideration before implementation.

Functionality wise the main TODOs are implementing more transaction
tests, e.g. multi input, segwit transaction, transactions with all
output types, groestl coin sighashes, cashaddr support and relay
protection tx support.

sedited committed:
Implement multi coin support interfaces

Extends the interfaces to take a coin argument. A coin contains a
cryptocurrency's constants. It also contains the name of a currency to
allow us to check for rule exceptions later on.

Also contains some fixes from review, such as splitting up some switch
cases, splitting up the tests and simplifying the Coin to bitcoinjs-lib
constant mapping function.

sedited committed:
Use RBF flag in create tx arguments instead of nsequence

Instead of passing in the raw sequence number, use a RBF flag to signal
if a transaction is replaceable. If it is, all transaction inputs will
have the RBF signalling nsequence number. Also remove the locktime from
the create tx arguments.

sedited committed:
Add WIF import and export functions

This removes the requirement of exposing bitcoin-js' ECPair class to
other files like the tests. Private keys can now be imported and
exported. Also introduces a small helper function to go from private to
public key. All keys are uncompressed.

sedited committed:
Implement Feedback from Edge Call

1. Throw errors to handle undefined types
2. Split up signing function
3. Add capability to guess address types
4. Private key buffer to hex
5. Code simplification around payment switch cases in address to script
hash converesion

sedited committed:
More transaction tests

sedited committed:
Change all buffers to strings in function interfaces

sedited committed:
Move code to its own utxobase/keymanager subdirectory

sedited committed:
Move coin constants to separate files

idlework commented on src/utxobased/keymanager/coinmapper.ts:
```diff
@@ -0,0 +1,13 @@
+const coinClasses: Coin[] = [new Litecoin(), new Bitcoin()]
+
+export function getCoinFromString(coinName: string): Coin {
+  const selectedCoin = coinClasses.filter(coin => coin.name === coinName)
```
the find function would be more readable than the filter function, and would avoid the !== 1 and array[0] implementation. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find

idlework reviewed (commented):
looks good. 1 Comment about the filter function in the getCoinFromString function.

sedited committed:
use find instead of filter in coinmapper

sedited committed:
Add bitcoin cash and cashaddr

This commit adds code to generate and parse a cashaddress.

The used implementation is a 'typescriptified' version of the old edge
wallet implementation, with some minor patches and improvements. It is
only called from the keymanager, which allows cashaddress use in the
same way as for other addresses.

To faciliate the larger number of tests and test the implementation of
the cashaddress generation directly and through the keymanager, the test
file is split up into multiple smaller files.

sedited committed:
More bitcoin transaction tests

Add transaction tests for bitcoin segwit and mix segwit and legacy
input transactions. There is still a bug prohibiting us from signing
p2wpkh-p2sh transaction inputs. These additionally require the redeem
script of the transaction. To this end the transaction input interface
will have to be slightly tweaked again.

sedited committed:
Add redeem script to transaction creation options

The redeem script is now returned by functions that go from public key
to its hash. The create transaction interfaces uses it to construct
wrapped segwit transactions.

sedited committed:
Use new directory structure, add edge's existing altcoin

sedited committed:
Add altcoin tests

The main TODOs left are:
1. Eboost test cases
2. Groestlcoin serialization and hashing implementation
3. Decred and Zcash long version bytes handling
4. UFO WIF key test case
5. Replay protection tx

sedited committed:
Virtual Transaction Size calculation, groestlcoin tx tests

The transaction virtual size is not estimated with a calculation during
tx creation.

sedited committed:
Move to fork of bitcoinjs-lib

This library now uses its own fork of the bitcoinjs lib. This allowed
implementation of bitcoincash and bitcoingold signatures, groestlcoin
signature hashing and more.

It also includes support for the legacy airbitz address format.

sedited committed:
Split p2sh generation from scripts

sedited committed:
Remove dependency on groestlcoinjs-lib

sedited committed:
Add declaration files for untyped modules

sedited committed:
Move decred bs58 functions

sedited committed:
Point bitcoinjslib to remote fork

sedited committed:
Add ZCash and Decred address derivation tests

sedited committed:
Fix smartcash support

sedited committed:
Remove replay protection from script types

Create a legacySeedToXPriv function that derives at m/0
Change the name from tx to psbt in signtx args
Remove p2wsh from script types
